UI: Add a connection info expander

jwijenbergh opened the pull request:
Fixes #248 

This commit adds a connection info expander to the UI.

Uses the NM api for querying, except where I could not find any
functions (e.g. rx and tx bytes, which are gathered from standard file
locations).

If it cannot find the data, as this is not foolproof, it will set it
to N/A, the application will keep working.

Note that glade changed the property names from `_` to `-`, but this
still works on the VMS. Apparently `_` are an old format that is not
used any more in glade.

Signed-off-by: jwijenbergh <jeroenwijenbergh@protonmail.com>

alvra commented:
Do you happen to know if getting these stats would also work for wireguard connections?

alvra commented on eduvpn/ui/stats.py:
```diff
@@ -0,0 +1,138 @@
+LINUX_STATS_FILE = "/sys/class/net/{iface}/statistics/{filename}"
+
+
+class NetworkStats():
```
It would be great to have a test of this, with `LINUX_STATS_FILE` mocked and checking that it parses correctly.

jwijenbergh replied:
great idea!

alvra commented on eduvpn/ui/stats.py:
```diff
@@ -0,0 +1,138 @@
+        """
+        hr_bytes = total_bytes
+        suffix = "TB"
+        for suffix in ["B", "KB", "MB", "GB"]:
```
nitpick, but technically it should be "kB"

jwijenbergh replied:
yeah makes sense :+1:

jwijenbergh commented:
> Do you happen to know if getting these stats would also work for wireguard connections?

They are normal networkmanager calls for getting the interface (which gives ip addresses) and the duration. The uploaded and downloaded stats are gathered from the standard `/sys/class/net/interface/statistics/{rx,tx}_bytes` files.

So the uploaded and downloaded stats work regardless of the protocol, the rest of the stats only work when networkmanager is used. So in short, if we integrate wireguard using networkmanager, these stats wll still be available. In this code I do not explicitly use any openvpn specific metrics.

gijzelaerr commented on eduvpn/nm.py:
```diff
@@ -44,6 +44,73 @@ def get_mainloop():
+    """
+    client = get_client()
+    active_connection = client.get_primary_connection()
+    if active_connection is None:
```
i would do `if not active_connection` everywhere where you compare to `is None`

jwijenbergh replied:
everywhere where I check for is None I only expect none and not other falsy types such as empty lists. But I will change it :+1:

gijzelaerr replied:
that is a good reason if things are possibly `0` or `[]`, but in this case not required.

gijzelaerr commented on eduvpn/ui/stats.py:
```diff
@@ -0,0 +1,138 @@
+        self.files = {}
+        self.start_bytes = {}
+        self.timestamp = None
+        self.default_text = "N/A"
```
i think we should use the translation helper function here

jwijenbergh replied:
I will try it! Does the rest of the ui translate too?

gijzelaerr replied:
yep!

gijzelaerr commented on eduvpn/ui/stats.py:
```diff
@@ -0,0 +1,138 @@
+
+logger = logging.getLogger(__name__)
+
+LINUX_STATS_FILE = "/sys/class/net/{iface}/statistics/{filename}"
```
did you try the `Path` from `pathlib`? that is quite useful and a bit more modern than manual path formatting.

jwijenbergh replied:
I didn't think it was really needed here as I'm just formatting the path once, but it does look neater

alvra commented on eduvpn/ui/ui.py:
```diff
@@ -593,8 +604,61 @@ def on_switch_connection_state(self, switch, state):
+            self.connection_info_stats.cleanup()
+            self.connection_info_stats = None
+
+    def start_connection_info(self, check_expanded=True):
```
Drop the argument and do the check separately when you need to.

gijzelaerr commented on eduvpn/ui/stats.py:
```diff
@@ -0,0 +1,138 @@
+            return None
+        return open(filepath, "r")
+
+    def get_human_readable_bytes(self, total_bytes):
```
we can move this method to utils, no need to find it to this class

jwijenbergh replied:
yep makes sense!

gijzelaerr commented on eduvpn/ui/stats.py:
```diff
@@ -0,0 +1,138 @@
+LINUX_STATS_FILE = "/sys/class/net/{iface}/statistics/{filename}"
+
+
+class NetworkStats():
```
the `()` are not required

gijzelaerr commented on eduvpn/ui/stats.py:
```diff
@@ -0,0 +1,138 @@
+        """
+        filepath = LINUX_STATS_FILE.format(iface=self.iface, filename=filename)
+        if not os.path.isfile(filepath):
+            logger.info(f"Network Stats: {filepath} is not a file")
```
i think this is a warning or even an error

jwijenbergh replied:
Yep!

gijzelaerr commented on eduvpn/ui/stats.py:
```diff
@@ -0,0 +1,138 @@
+        """
+        # If interface wasn't set
+        if self.iface is None:
+            self.iface = get_iface()
```
what happens if the iface changes? Does this code handle that properly?  If not, maybe not set the iface but just always get it

jwijenbergh replied:
I don't think the interface will ever change when the vpn connection is active right? This class is alive as long as the connection is

gijzelaerr replied:
ok if it logically makes sense then it's okay, but in general I think it is best to store as minimal state as possible unless it is inside a busy loop, which this is not.

jwijenbergh replied:
This gets run twice (once for upload and download) every second so I want to use as little resources as possible, this is especially the reason why I keep the files open and only close them after the thread is finished.

gijzelaerr replied:
ah right, thats true. maybe it should be set only once during a state transition as described in the other thread.

gijzelaerr commented on eduvpn/ui/stats.py:
```diff
@@ -0,0 +1,139 @@
+
+        return f"{hr_bytes:.2f} {suffix}"
+
+    def get_file_bytes(self, filename, typename):
```
this can use some argument type annotations

jwijenbergh replied:
yeah makes sense

alvra commented on eduvpn/ui/ui.py:
```diff
@@ -593,8 +604,61 @@ def on_switch_connection_state(self, switch, state):
+            if not is_expanded:
+                return
+
+        if not self.app.network_state.has_transition('disconnect'):
```
If you don't show based on network state here, will the stats show when the network state changes back? I think it's better to just show the stats regardless of network state, however.

jwijenbergh replied:
Do you mean that I should show the connection stats when a VPN is not active?

jwijenbergh replied:
You can reopen this if this is not right what I just pushed, but I think it does make sense as an extra error check

gijzelaerr commented on eduvpn/ui/stats.py:
```diff
@@ -0,0 +1,139 @@
+
+        # Get the statistic from the file
+        # and go to the beginning
+        stat = float(current_file.readline())
```
i think we should move the opening and parsing of the file to a separate helper function, this method is getting quite long.

jwijenbergh replied:
makes sense!

alvra commented on eduvpn/ui/stats.py:
```diff
@@ -0,0 +1,139 @@
+class NetworkStats():
+    def __init__(self, interval):
+        self.iface = None
+        self.interval = interval
```
This doesn't seem to be used.

jwijenbergh replied:
ah yeah sorry the interval was for the /s data rate, that is not needed any more. Thanks

gijzelaerr commented on eduvpn/ui/stats.py:
```diff
@@ -0,0 +1,139 @@
+        """
+        return self.get_file_bytes("tx_bytes", "upload")
+
+    def get_ip_address(self, name, func):
```
adding a Callable type annotation makes it easier to read

gijzelaerr commented:
Great work! Thanks, @jwijenbergh. I have some small comments, but they are mostly esthetic. One thing though, Although it is reasonably self-contained I think the `NetworkStats` might be a bit too complex for what it is doing. It might help to split it up in a couple of separate functions, unless a method needs to have access to a local state there is no need to bundle it with a class. That makes it easier to test also which brings me to the next point, i'm missing unit tests :)

alvra reviewed (commented):
Look good! Haven't looked at the NM code in detail or tested though, but nice to get this working

jwijenbergh commented on eduvpn/ui/stats.py:
```diff
@@ -0,0 +1,139 @@
+        """
+        address = self.addresses.get(name, None)
+        if address is None:
+            address = func()
```
because then we get duplicated code, for ipv4 and ipv6

jwijenbergh commented on eduvpn/ui/app.py:
```diff
@@ -129,6 +129,7 @@ def enter_ConnectedState(self, old_state, new_state):
         self.connection_notification.show(
             title=_("Connected"),
             message=_("You are now connected to your server."))
+        self.window.start_connection_info()
```
The reason I added it here is that there doesn't seem to be a good way to check in ui.py if the connection is *actually* active. When I did it in ui.py in the connection switch I could get a bad state as we're starting the stats too early

gijzelaerr replied:
i might be wrong, but i guess the answer there is to have a callback mechanism that is triggered when the state is changed and then check if that state is ready. then you are sure you have a valid state. @alvra might already have state logic for that.

alvra replied:
You could define the same transition callback for on window, but it's better to call the start/stop in the enter/exit for a single transitions because its easier to check and maintain that it's correct. In your case you might need one more: enter/exit ConnectionStatus and when toggling the stats itself.

jwijenbergh replied:
Yeah that is the state transition here where I added it, I could add the same state transition to ui?

jwijenbergh replied:
> You could define the same transition callback for on window, but it's better to call the start/stop in the enter/exit for a single transitions because its easier to check and maintain that it's correct. In your case you might need one more: enter/exit ConnectionStatus and when toggling the stats itself.

whoops didn't see this, so should the stats toggling also be a separate state transition?

jwijenbergh commented on eduvpn/ui/stats.py:
```diff
@@ -0,0 +1,138 @@
+        Helper function to calculate the human readable bytes.
+        E.g. B, KB, MB, GB, TB.
+        """
+        hr_bytes = total_bytes
```
Do you mean that specific line or the function?

alvra replied:
The line, renaming the variable.

jwijenbergh replied:
ah it's indeed not needed but it gives the impression that we're in the end converting it to human readable bytes, modifying the argument directly inside the function makes it more vague IMO

jwijenbergh commented:
> Great work! Thanks, @jwijenbergh. I have some small comments, but they are mostly esthetic. One thing though, Although it is reasonably self-contained I think the `NetworkStats` might be a bit too complex for what it is doing. It might help to split it up in a couple of separate functions, unless a method needs to have access to a local state there is no need to bundle it with a class. That makes it easier to test also which brings me to the next point, i'm missing unit tests :)

Yeah makes sense, tests for networkstats are needed

jwijenbergh commented:
Steps done:

- Fixed reviews, except for the ones I left open because I am waiting for follow up
- Cleaned up networkstats, I used lru cache on properties as that made the code a lot cleaner
- Added tests for upload and download (rest is not possible as that uses the NM api)

jwijenbergh commented:
Still needs some fedora fixes as that apparently does not copy over the test files to the correct directory

gijzelaerr commented on eduvpn/ui/stats.py:
```diff
@@ -0,0 +1,146 @@
+
+    # These properties define an LRU cache
+    # This cache is used so that we do not query these every second
+    @property  # type: ignore
```
why do these types need to be ignored?

jwijenbergh replied:
Because mypy doesn't really like cached properties, for that you need ```@functools.cached_property```, but that is only available for python 3.8 and up

jwijenbergh replied:
see https://github.com/python/mypy/issues/1362

gijzelaerr commented on eduvpn/ui/stats.py:
```diff
@@ -0,0 +1,146 @@
+            return None
+        return open(filepath, "r")
+
+    def get_file_bytes(self, filehandler):
```
filehandler needs a type annotation

gijzelaerr commented on eduvpn/utils.py:
```diff
@@ -169,3 +169,18 @@ def parse_http_date_header(date: str) -> datetime:
 parse_http_expires_header = parse_http_date_header
+
+
+def get_human_readable_bytes(total_bytes):
```
type annotation

gijzelaerr commented on tests/data/stats/mock/statistics/tx_bytes_increased_big:
```diff
@@ -0,0 +1 @@
+61224753204
```
this solution is not bad, but i think it might be better to create these mock files on the fly in a TemporaryDirectory:

https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory

So request a temp folder during mocking, write some files for the test case and then point the logic to read those files. Don't forget to clean up afterwards. This way we dont need to worry about all these temp files for special use cases.

Thing is, we otherwise need to bundle these files also with the tarball, which requires updating the `MANIFEST.in` file. Otherwise, the files are not bundled with the source tarball and the eventual packages fail to build. I prefer to keep this kind of admin minimal, and generating temp files is cleaner (as long as you remove the temp files afterwards).

You can also create a Mock IO object per `open()`, but that might be more work.

jwijenbergh replied:
Thanks that makes sense!

gijzelaerr commented on eduvpn/ui/ui.py:
```diff
@@ -511,6 +528,10 @@ def context_ConnectionStatus(self, state):
             'update-validity',
         ))
 
+    @transition_edge_callback(ENTER, network_state.DisconnectedState)
```
I leave this to @alvra to verify if this logic is sound.

alvra commented on eduvpn/ui/ui.py:
```diff
@@ -511,6 +528,10 @@ def context_ConnectionStatus(self, state):
             'update-validity',
         ))
 
+    @transition_edge_callback(ENTER, network_state.DisconnectedState)
```
And use the correct function name for this transition.

jwijenbergh replied:
> Keep enter/exit transitions for the same state together.

This is the first use of the disconnectedstate

jwijenbergh replied:
> And use the correct function name for this transition.

This is entering the disconnected state, so `enter_DisconnectedState` makes sense to me

alvra commented on eduvpn/ui/ui.py:
```diff
@@ -493,6 +503,12 @@ def enter_ConfiguringConnection(self, old_state, new_state):
 
+    @transition_edge_callback(ENTER, network_state.ConnectedState)
+    def enter_ConnectedState(self, old_state, new_state):
+        is_expanded = self.connection_info_expander.get_expanded()
```
Maybe the function should be named `is_expanded` if you feel the need to use a named variable here.

jwijenbergh replied:
I cannot change this, this is the gtk api

alvra replied:
Yes, sorry

jwijenbergh replied:
No worries!

alvra reviewed (commented):
The issue I have with starting/stopping the stats collector is not one of correctness but simplicity, and ultimately correctness and maintainability. Let me try to explain.

As far as I can tell, you're trying it to run when 3 conditions are met, handled in two different ways.
  * The interface is showing the connection status page
  * The stats toggle is on
  * The network is "connected"

From reading the code, it seems like you're missing at least one option: leaving the conn status page with the stats on, and then going back to it; the toggle would still be on. You could fix this one case by adding this transition function, but that only makes the logic more complex.

The start/stop logic is now spread all over the UI class, which makes these things hard to verify. Another way to see this is that you basically have to copy the same start/stop handler each 3 times.

I don't know what the best tradeoff is here, this will depend on the situation that you probably know best.
  * Is the stats collecting that expensive that it must absolutely only run when needed?
  * What are the UX consequences when not collecting in some cases, and showing the stats as N/A?

jwijenbergh commented on tests/test_stats.py:
```diff
@@ -0,0 +1,55 @@
+    @patch('eduvpn.ui.stats.NetworkStats.iface', MOCK_IFACE)
+    def test_upload(self):
+        class_ = stats.NetworkStats()
+        self.assertEqual(class_.download, "0.00 B")
```
Agreed

jwijenbergh commented on eduvpn/utils.py:
```diff
@@ -169,3 +169,18 @@ def parse_http_date_header(date: str) -> datetime:
+    E.g. B, KB, MB, GB, TB.
+    """
+    suffix = ""
+    for suffix in ["B", "kB", "mB", "gB", "tB"]:
```
Ah okay, apologies

jwijenbergh commented on eduvpn/ui/stats.py:
```diff
@@ -0,0 +1,146 @@
+    # These properties define an LRU cache
+    # This cache is used so that we do not query these every second
+    @property  # type: ignore
+    @lru_cache(maxsize=1)
```
that uses a maxsize=None, I want to explicitly mention that the maxsize=1, this is done everywhere throughout the code as well. See nm.py

alvra replied:
Sure, but does it matter?

jwijenbergh replied:
I get your point but @cache is only used once AFAICT whereas @lru_cache is used much more often. We would have to convert others in a separate pr as well then

jwijenbergh replied:
I have changed it, thanks :+1:. I might make a pr later to convert the rest

jwijenbergh commented on eduvpn/ui/stats.py:
```diff
@@ -0,0 +1,146 @@
+import logging
+import time
+from datetime import timedelta
+from gettext import gettext
```
Didn't see it here so I thought it was fine https://github.com/eduvpn/python-eduvpn-client/blob/42dee5922e8df73d8d177eed511c190a5e6d0df4/eduvpn/network.py#L6

alvra replied:
But there it's not used to directly mark a translated string.

Actually you can use the `translated_property` from there (if you move it to utils).

jwijenbergh commented on eduvpn/ui/stats.py:
```diff
@@ -0,0 +1,146 @@
+        if file_bytes_download is None:
+            return self.default_text
+        if file_bytes_download <= self.start_bytes_download:
+            return get_human_readable_bytes(0.0)
```
I have fixed this already, will push my changes later

jwijenbergh commented:
> The issue I have with starting/stopping the stats collector is not one of correctness but simplicity, and ultimately correctness and maintainability. Let me try to explain.
> 
> As far as I can tell, you're trying it to run when 3 conditions are met, handled in two different ways.
> 
>     * The interface is showing the connection status page
> 
>     * The stats toggle is on
> 
>     * The network is "connected"
> 
> 
> From reading the code, it seems like you're missing at least one option: leaving the conn status page with the stats on, and then going back to it; the toggle would still be on. You could fix this one case by adding this transition function, but that only makes the logic more complex.
> 
> The start/stop logic is now spread all over the UI class, which makes these things hard to verify. Another way to see this is that you basically have to copy the same start/stop handler each 3 times.
> 
> I don't know what the best tradeoff is here, this will depend on the situation that you probably know best.
> 
>     * Is the stats collecting that expensive that it must absolutely only run when needed?
> 
>     * What are the UX consequences when not collecting in some cases, and showing the stats as N/A?

I agree, but for simplicity I left it the way it is right now. The android app also collapses the connection info when you go back. And yes I do want to run the stats collecting as little as possible. I do not want an app to query the connection info every second when it is collapsed.

alvra commented:
Won't the GTK toggle still be on? I don't think its reset when switching pages.

jwijenbergh commented:
> Won't the GTK toggle still be on? I don't think its reset when switching pages.

Apologies, this is actually handled differently. It is kept open. So if you go back and enter again it will still gather data with the popup open. Do we want to explicitly collapse this as we go back? I made sure that no data gets collected when you go back.

On the android app you can only go back when you're in the disconnected state, in the connected state there is not really a back button.

jwijenbergh commented:
Steps taken:

- Added typing annotations everywhere
- Cleaned up floats and ints inconsistencies
- Now a temp dir gets created instead of having test data
- Fixed most comments by reviewers, except the ones that are left open
